api/v3/Case.php

```diff
@@ -60,37 +60,67 @@
  *   api result array
  */
 function civicrm_api3_case_create($params) {
+  // Process params
+  $values = array();
+  _civicrm_api3_custom_format_params($params, $values, 'Case');
+  $params = array_merge($params, $values);
 
-  if (!empty($params['id'])) {
-    return civicrm_api3_case_update($params);
-  }
-
-  civicrm_api3_verify_mandatory($params, NULL, array(
-    'contact_id',
-    'subject',
-    array('case_type', 'case_type_id'))
-  );
   _civicrm_api3_case_format_params($params);
 
-  // If format_params didn't find what it was looking for, return error
-  if (empty($params['case_type_id'])) {
-    throw new API_Exception('Invalid case_type. No such case type exists.');
-  }
-  if (empty($params['case_type'])) {
-    throw new API_Exception('Invalid case_type_id. No such case type exists.');
+  if (empty($params['id'])) {
+    // Creating a new case, so make sure we have the necessary parameters
+    civicrm_api3_verify_mandatory($params, NULL, [
+        'contact_id',
+        'subject',
+        ['case_type', 'case_type_id']
+      ]
+    );
   }
+  else {
+    // Update an existing case
+    // FIXME: Some of this logic should move to the BAO object?
+    // FIXME: Should we check if case with ID actually exists?
+    if (!isset($params['case_id']) && isset($params['id'])) {
+      $params['case_id'] = $params['id'];
+    }
 
-  // Fixme: can we safely pass raw params to the BAO?
-  $newParams = array(
-    'case_type_id' => $params['case_type_id'],
-    'creator_id' => $params['creator_id'],
-    'status_id' => $params['status_id'],
-    'start_date' => $params['start_date'],
-    'end_date' => CRM_Utils_Array::value('end_date', $params),
-    'subject' => $params['subject'],
-  );
+    // return error if modifying creator id
+    if (array_key_exists('creator_id', $params)) {
+      throw new API_Exception(ts('You cannot update creator id'));
+    }
+
+    $mCaseId = $origContactIds = array();
+
+    // get original contact id and creator id of case
+    if (!empty($params['contact_id'])) {
+      $origContactIds = CRM_Case_BAO_Case::retrieveContactIdsByCaseId($params['id']);
+      $origContactId = $origContactIds[1];
+    }
+
+    if (count($origContactIds) > 1) {
+      // check valid orig contact id
+      if (!empty($params['orig_contact_id']) && !in_array($params['orig_contact_id'], $origContactIds)) {
+        throw new API_Exception('Invalid case contact id (orig_contact_id)');
+      }
+      elseif (empty($params['orig_contact_id'])) {
+        throw new API_Exception('Case is linked with more than one contact id. Provide the required params orig_contact_id to be replaced');
+      }
+      $origContactId = $params['orig_contact_id'];
+    }
 
-  $caseBAO = CRM_Case_BAO_Case::create($newParams);
+    // check for same contact id for edit Client
```

Comment:
Perhaps it would be better to use the comment to explain why do we need this check and the merge?